patch by Ishimoto Shinobu
based on Szabolcs Nagy's work

in gcc-9 -static means static non-pie, even if -pie or -static-pie
are specified, -static-pie can be used to build static pie.

in musl toolchains -static -pie always meant static pie, so this
patch fixes the link specs accordingly, the new -static-pie is just
an alias to -static -pie.

МЫ ЕБАЛИСЬ ХУИ СЛОМАЛИСЬ
И ВСЕ ЖОПЫ ОБ НИХ ПОРВАЛИСЬ
ТОЛЬКО ВСЕ РАВНО СЕКСА МАЛО
МЫ ХОТИМ ЧТОБ ТЫ СКАЗАЛА

отъеби меня скорей
своих сил ты не желей
и засунь поглубже мне
18 ведь уже

diff -ruN gcc-9.1.0-orig/gcc/common.opt gcc-9.1.0/gcc/common.opt
--- gcc-9.1.0-orig/gcc/common.opt	2019-03-09 18:28:04.000000000 +0900
+++ gcc-9.1.0/gcc/common.opt	2019-05-03 21:36:03.327947229 +0900
@@ -3287,11 +3287,11 @@
 
 no-pie
 Driver RejectNegative Negative(shared)
-Don't create a dynamically linked position independent executable.
+Don't create a position independent executable.
 
 pie
 Driver RejectNegative Negative(no-pie)
-Create a dynamically linked position independent executable.
+Create a position independent executable.
 
 static-pie
 Driver RejectNegative Negative(pie)
diff -ruN gcc-9.1.0-orig/gcc/config/gnu-user.h gcc-9.1.0/gcc/config/gnu-user.h
--- gcc-9.1.0-orig/gcc/config/gnu-user.h	2019-01-16 18:38:21.000000000 +0900
+++ gcc-9.1.0/gcc/config/gnu-user.h	2019-05-03 21:38:10.904937016 +0900
@@ -51,13 +51,12 @@
 #define GNU_USER_TARGET_STARTFILE_SPEC \
   "%{shared:; \
      pg|p|profile:%{static-pie:grcrt1.o%s;:gcrt1.o%s}; \
-     static:crt1.o%s; \
-     static-pie:rcrt1.o%s; \
+     static|static-pie:%{" PIE_SPEC ":rcrt1.o%s;:crt1.o%s}; \
      " PIE_SPEC ":Scrt1.o%s; \
      :crt1.o%s} " \
    GNU_USER_TARGET_CRTI " \
-   %{static:crtbeginT.o%s; \
-     shared|static-pie|" PIE_SPEC ":crtbeginS.o%s; \
+   %{shared|" PIE_SPEC ":crtbeginS.o%s; \
+     static:crtbeginT.o%s; \
      :crtbegin.o%s} \
    %{fvtable-verify=none:%s; \
      fvtable-verify=preinit:vtv_start_preinit.o%s; \
@@ -76,8 +75,7 @@
   "%{fvtable-verify=none:%s; \
      fvtable-verify=preinit:vtv_end_preinit.o%s; \
      fvtable-verify=std:vtv_end.o%s} \
-   %{static:crtend.o%s; \
-     shared|static-pie|" PIE_SPEC ":crtendS.o%s; \
+   %{shared|" PIE_SPEC ":crtendS.o%s; \
      :crtend.o%s} " \
    GNU_USER_TARGET_CRTN " " \
    CRTOFFLOADEND
@@ -106,7 +104,7 @@
 #define LIB_SPEC GNU_USER_TARGET_LIB_SPEC
 
 #if defined(HAVE_LD_EH_FRAME_HDR)
-#define LINK_EH_SPEC "%{!static|static-pie:--eh-frame-hdr} "
+#define LINK_EH_SPEC "%{!static|" PIE_SPEC ":--eh-frame-hdr} "
 #endif
 
 #define GNU_USER_TARGET_LINK_GCC_C_SEQUENCE_SPEC \
diff -ruN gcc-9.1.0-orig/gcc/gcc.c gcc-9.1.0/gcc/gcc.c
--- gcc-9.1.0-orig/gcc/gcc.c	2019-03-11 22:58:44.000000000 +0900
+++ gcc-9.1.0/gcc/gcc.c	2019-05-03 21:41:00.976965660 +0900
@@ -931,12 +931,12 @@
 #ifndef LINK_PIE_SPEC
 #ifdef HAVE_LD_PIE
 #ifndef LD_PIE_SPEC
-#define LD_PIE_SPEC "-pie"
+#define LD_PIE_SPEC "-pie %{static|static-pie:--no-dynamic-linker -z text -Bsymbolic}"
 #endif
 #else
 #define LD_PIE_SPEC ""
 #endif
-#define LINK_PIE_SPEC "%{static|shared|r:;" PIE_SPEC ":" LD_PIE_SPEC "} "
+#define LINK_PIE_SPEC "%{shared|r:;" PIE_SPEC ":" LD_PIE_SPEC "} "
 #endif
 
 #ifndef LINK_BUILDID_SPEC
